### PR TITLE
Remove Settings.getAsClass

### LINKED ISSUE #12643: Unable to start AWS discovery with 2.0.0-SNAPSHOT
I started 2 AWS instances using latest 2.0.0-SNAPSHOT I built locally.

When starting with defaults, everything is fine.

When changing `elasticsearch.yml` to:

```yml
discovery.type: ec2
discovery.ec2.tag.Name: dadoonet-test-2.0.0-SNAP
```

I get this error while launching:

```
[2015-08-04 16:37:18,363][ERROR][org.elasticsearch.bootstrap] Exception
NoClassSettingsException[Failed to load class setting [discovery.type] with value [ec2]]; nested: ClassNotFoundException[org.elasticsearch.discovery.ec2.Ec2DiscoveryModule];
	at org.elasticsearch.common.settings.Settings.loadClass(Settings.java:604)
	at org.elasticsearch.common.settings.Settings.getAsClass(Settings.java:592)
	at org.elasticsearch.discovery.DiscoveryModule.spawnModules(DiscoveryModule.java:53)
	at org.elasticsearch.common.inject.ModulesBuilder.add(ModulesBuilder.java:44)
	at org.elasticsearch.node.Node.<init>(Node.java:177)
	at org.elasticsearch.node.NodeBuilder.build(NodeBuilder.java:157)
	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:177)
	at org.elasticsearch.bootstrap.Bootstrap.main(Bootstrap.java:272)
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:28)
Caused by: java.lang.ClassNotFoundException: org.elasticsearch.discovery.ec2.Ec2DiscoveryModule
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	at org.elasticsearch.common.settings.Settings.loadClass(Settings.java:602)
	... 8 more
```

The cloud plugin contains needed libs I think:

```
[ec2-user@ip-10-0-0-113 cloud-aws]$ ll
total 5776
-rw-rw-r-- 1 ec2-user ec2-user  513834  4 août  16:32 aws-java-sdk-core-1.10.0.jar
-rw-rw-r-- 1 ec2-user ec2-user 2153751  4 août  16:32 aws-java-sdk-ec2-1.10.0.jar
-rw-rw-r-- 1 ec2-user ec2-user  258130  4 août  16:32 aws-java-sdk-kms-1.10.0.jar
-rw-rw-r-- 1 ec2-user ec2-user  563917  4 août  16:32 aws-java-sdk-s3-1.10.0.jar
-rw-rw-r-- 1 ec2-user ec2-user  232771  4 août  16:32 commons-codec-1.6.jar
-rw-rw-r-- 1 ec2-user ec2-user   62050  4 août  16:32 commons-logging-1.1.3.jar
-rw-rw-r-- 1 ec2-user ec2-user   44814  4 août  16:32 elasticsearch-cloud-aws-2.0.0-SNAPSHOT.jar
-rw-rw-r-- 1 ec2-user ec2-user  592008  4 août  16:32 httpclient-4.3.6.jar
-rw-rw-r-- 1 ec2-user ec2-user  282793  4 août  16:32 httpcore-4.3.3.jar
-rw-rw-r-- 1 ec2-user ec2-user   39815  4 août  16:32 jackson-annotations-2.5.0.jar
-rw-rw-r-- 1 ec2-user ec2-user 1143162  4 août  16:32 jackson-databind-2.5.3.jar
-rw-rw-r-- 1 ec2-user ec2-user    2253  4 août  16:32 plugin-descriptor.properties
```

And the cloud jar file contains the class we are looking for: `org/elasticsearch/discovery/ec2/Ec2DiscoveryModule.class`

```
[ec2-user@ip-10-0-0-113 cloud-aws]$ unzip -l elasticsearch-cloud-aws-2.0.0-SNAPSHOT.jar 
Archive:  elasticsearch-cloud-aws-2.0.0-SNAPSHOT.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  08-04-2015 17:40   META-INF/
      525  08-04-2015 17:40   META-INF/MANIFEST.MF
        0  08-04-2015 17:40   org/
        0  08-04-2015 17:40   org/elasticsearch/
        0  08-04-2015 17:40   org/elasticsearch/cloud/
        0  08-04-2015 17:40   org/elasticsearch/cloud/aws/
        0  08-04-2015 17:40   org/elasticsearch/cloud/aws/blobstore/
        0  08-04-2015 17:40   org/elasticsearch/cloud/aws/network/
        0  08-04-2015 17:40   org/elasticsearch/cloud/aws/node/
        0  08-04-2015 17:40   org/elasticsearch/discovery/
        0  08-04-2015 17:40   org/elasticsearch/discovery/ec2/
        0  08-04-2015 17:40   org/elasticsearch/plugin/
        0  08-04-2015 17:40   org/elasticsearch/plugin/cloud/
        0  08-04-2015 17:40   org/elasticsearch/plugin/cloud/aws/
        0  08-04-2015 17:40   org/elasticsearch/repositories/
        0  08-04-2015 17:40   org/elasticsearch/repositories/s3/
     8232  08-04-2015 17:40   org/elasticsearch/cloud/aws/AwsEc2Service.class
     1641  08-04-2015 17:40   org/elasticsearch/cloud/aws/AwsModule.class
      689  08-04-2015 17:40   org/elasticsearch/cloud/aws/AwsS3Service.class
     1232  08-04-2015 17:40   org/elasticsearch/cloud/aws/AwsSigner.class
    11319  08-04-2015 17:40   org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.class
     7517  08-04-2015 17:40   org/elasticsearch/cloud/aws/blobstore/S3BlobContainer.class
     6371  08-04-2015 17:40   org/elasticsearch/cloud/aws/blobstore/S3BlobStore.class
     3246  08-04-2015 17:40   org/elasticsearch/cloud/aws/blobstore/S3OutputStream.class
     8992  08-04-2015 17:40   org/elasticsearch/cloud/aws/InternalAwsS3Service.class
     2076  08-04-2015 17:40   org/elasticsearch/cloud/aws/network/Ec2NameResolver$Ec2HostnameType.class
     4004  08-04-2015 17:40   org/elasticsearch/cloud/aws/network/Ec2NameResolver.class
     3668  08-04-2015 17:40   org/elasticsearch/cloud/aws/node/Ec2CustomNodeAttributes.class
     1139  08-04-2015 17:40   org/elasticsearch/discovery/ec2/AwsEc2UnicastHostsProvider$1.class
     1472  08-04-2015 17:40   org/elasticsearch/discovery/ec2/AwsEc2UnicastHostsProvider$HostType.class
     9807  08-04-2015 17:40   org/elasticsearch/discovery/ec2/AwsEc2UnicastHostsProvider.class
     2317  08-04-2015 17:40   org/elasticsearch/discovery/ec2/Ec2Discovery.class
     1635  08-04-2015 17:40   org/elasticsearch/discovery/ec2/Ec2DiscoveryModule.class
     2631  08-04-2015 17:40   org/elasticsearch/plugin/cloud/aws/CloudAwsPlugin.class
     6781  08-04-2015 17:40   org/elasticsearch/repositories/s3/S3Repository.class
     1099  08-04-2015 17:40   org/elasticsearch/repositories/s3/S3RepositoryModule.class
        0  08-04-2015 17:40   META-INF/maven/
        0  08-04-2015 17:40   META-INF/maven/org.elasticsearch.plugin/
        0  08-04-2015 17:40   META-INF/maven/org.elasticsearch.plugin/elasticsearch-cloud-aws/
     2172  08-04-2015 16:39   META-INF/maven/org.elasticsearch.plugin/elasticsearch-cloud-aws/pom.xml
      142  08-04-2015 17:40   META-INF/maven/org.elasticsearch.plugin/elasticsearch-cloud-aws/pom.properties
---------                     -------
    88707                     41 files
```

### LINKED ISSUE #12656: Plugin entry points using Settings.getAsClass are broken
I think the underlying issue in #12643 is how `SpawnModules` works. This is an entry point for various pluggable classes that goes out and tries to load plugged in modules from the classloader. However, this no longer works now that plugins get their own classloader.

I think the right thing to do is eliminate SpawnModules altogether. Any plugin entry points should:
1. Have a method on a module the plugin can register new classes with
2. Have tests that these entry points actually work with a class *outside of the core classloader*

I think this has to block 2.0 beta, otherwise many plugins are just broken (like all the discovery plugins right now).

### PR DESCRIPTION
**Pull request by rjernst**
This method on settings loaded a class, based on a setting value, using the default classloader. It had all kinds of leniency in how the classname was found, and simply cannot work with plugins having isolated classloaders.

This change removes that method. Some of the uses of it were for custom extension points, like custom repository or discovery types. A lot were just there to plugin mock implementations for tests. For the settings that were legitimate, all now support plugins adding the given setting via onModule. For those that were specific to tests for mocks, they now use Classes.loadClass (a helper around Class.forName). This is a temporary measure until (in a future PR) tests can change the implementation via package private statics.

I also removed a number of unnecessary intermediate modules, added a "jvm-example" plugin that can be filled in in the future as a smoke test for breaking plugins, and gave some documentation to "spawn" modules interface.

closes #12643
closes #12656

### CODE REVIEW

**Comment by dadoonet** [comment]
That's fantastic!
I just looked at the plugin part and it looks great to me.

**Comment by rmuir** [comment]
+1

File: core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java [hunk 1]
```diff
@@ -20,9 +20,19 @@
...
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
  */
-public interface SpawnModules {
+@Deprecated
+public interface SpawnShittyModules {
```
Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: sorry... -1 revert the name

File: core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java [hunk 2]
```diff
@@ -20,9 +20,19 @@
...
+ * This interface can be added to a Module to spawn sub modules. DO NOT USE.
  *
+ * This is fundamentally broken.
+ * <ul>
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
```
Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: this won't play along with module pre processing... so creating modules directly will actually be broken

File: core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java [hunk 3]
```diff
@@ -20,9 +20,19 @@
 package org.elasticsearch.common.inject;
 
 /**
+ * This interface can be added to a Module to spawn sub modules. DO NOT USE.
  *
+ * This is fundamentally broken.
+ * <ul>
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
```
Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: agree... don't use `SpawnModules` for extensibility... extensibility is not the goal of this interface. The goal of this interface is modularization (enabling modules to be composed of other modules) and to play along with the module pre processing phase.

File: core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java [hunk 4]
```diff
@@ -20,9 +20,19 @@
...
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
  */
-public interface SpawnModules {
+@Deprecated
```
Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: -1 on deprecation at this point, as the logic behind it is broken

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: The logic is not broken. This interface is broken, for the reasons described in the javadoc. All this interface does is make for (1) more classes (2) more obfuscation and (3) possibility for broken pluggability APIs.  If this interface stays around, nothing stops new extension points from being added which are broken again by trying to plugin a module "implementation", which as explained, simply cannot work with plugin isolation.

I don't know what you mean by "module pre processing", but extension points should be clear, and plugins should bind ancillary classes they need in their own modules. For example, if a plugin wants to bind a custom transport protocol, it should add the transport protocol to TransportModule with onModule, and any ancillary classes it needs bound can be added in its own module. There is no need to "group" these bindings together in one module.

Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: > (1) more classes

sure, I can leave with a proposed alternative that will remove the class, but quite frankly, I don't care if we have an extra class here

> (2) more obfuscation

note sure what's obfuscated here... it's as simple as "modules that implement this depend on other modules"

>  (3) possibility for broken pluggability APIs

Give me any extensible module that doesn't implement this interface and I'll show you how I can implement it in a broken way. As I mentioned, this interface is not for extensibility, but for modularity. Abusing things can be done everywhere... and we need to make sure that we carefully define the contract for APIs. And documentation **is** part of the contract - I'll be happy to see an explanation in the javadoc that states 1) what this interface is for and all about in general, 2) that you should not abuse it for extensibility and if you do want extensibility point to a top package level javadoc that explains what is the right way of opening an extension point.

> I don't know what you mean by "module pre processing"

https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/inject/PreProcessModule.java

> but extension points should be clear, and plugins should bind ancillary classes they need in their own modules.

totally agree, and that's also the case with the `SpanwModules` itself. The only thing that happens here in addition is the ability to group modules per functionality/layer if you will. To your example, if I'm replacing the `TransportModule`, I don't want to declare that I'm replacing it in one place and in another place define the sub-modules of the new transport module. I want to simply declare "this is the new transport module" and it comes with whatever sub-modules that it defines - one place - clean. On the same note, if you choose to remove/replace a module, I don't want to think carefully what other modules I need to replace along with it - that what you'll end up with if you flatten all the modules - there're no clear module boundaries and dependencies. 

The problem that you're referring to in the javadoc (and I generally agree that there is a problem here) should be fixed, but in the appropriate place and the spawn modules should not be "blamed" for it. The problem here is that there needs to be a well defined function for replacing modules, on the container level.. somewhere you can simply call:

```
replace(TransportModule.class).with(MyTransportModule.class)
```

and the implementation should take care of unregistering the sub-modules of the prev. one and registering the sub-modules of the new one. 

Removing this interface will prevent defining well define & separated modules and promote module spaghetti - where every module is a top level module... without proper modularisation.

I'd remove this change from this PR and I'm all for considering another PR that really tackles the problem here, providing a good alternative that still promotes modularity. Deprecating this interface does not provide a better/safer solution, but instead forces to flatten all modules across the board, which is really bad IMO.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: > I'd remove this change from this PR and I'm all for considering another PR that really tackles the problem here, providing a good alternative that still promotes modularity.

uh, when you find a crappy/troublesome api or functionality, there need not always be a replacement. Sometimes things are just a bad idea and should be removed completely.

Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: > uh, when you find a crappy/troublesome api or functionality, there need not always be a replacement. Sometimes things are just a bad idea and should be removed completely.

I wholeheartedly agree with you. And to add to that, if removing the crappy api introduces another crappy thing (api/design/etc..), then you need to rethink the change and sometimes avoid it until you come up with a better alternative (and that's where I believe we're standing on this issue)

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: > if removing the crappy api introduces another crappy thing (api/design/etc..)

But that's not the case here. Flattening modules is a good thing. It will make dependencies of services/components much less "spaghetti". Obviously I can't do flattening/cleanup in this PR. But the deprecation is here to warn against adding more of this craziness.

> The problem here is that there needs to be a well defined function for replacing modules, on the container level..

I disagree. There should be no need to replace a module, there is no need to make every part of Elasticsearch pluggable. We should have very specific pieces that are overrideable. Using the DiscoveryModule as an example, we have "local" and "zen" discovery types that are "builtin" to ES. These are always available, and zen has a number of associated classes that are bound in discovery module. However, if a plugin wants to add a new discovery type, the Discovery class is the only thing they are plugging in, and that is registerable with the DiscoveryModule. If they require extra bindings, they can do so in their own module, which is added as part of their plugin. All of our extension points should work this way.

Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: I'll be very happy if we find ourselves in a place where we simply don't need to replace a module. I didn't check the full code base to validate that indeed that works. I just know that I had to extend a module once for testability (i.e. exposing the extensibility for tests only, without exposing the extensibility for plugins). 

In any case, that doesn't change the reasoning behind properly modularizing the code base such that, 1) each module is the only entity responsible for its sub modules, 2) modules can be extensible at any level

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: > I wholeheartedly agree with you. And to add to that, if removing the crappy api introduces another crappy thing (api/design/etc..), then you need to rethink the change and sometimes avoid it until you come up with a better alternative (and that's where I believe we're standing on this issue)

But its not removed here, only deprecated. You just have sour grapes because we are telling you your code fucking sucks. Thats fine, I'm -1 to push this fix *without* the deprecation so we will figure this shit out. Great because its a blocker issue.

And guess what: I won't back down. Ever.

Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: > But its not removed here, only deprecated.

deprecation is not the right mechanism as IMO we're deprecating the wrong thing here. It'd be like me deprecating the `Engine` just because there's an issue with index recovery. We need a separate issue for this... and sure, label it for 2.0... but it's not a blocker for beta1.

> You just have sour grapes because we are telling you your code fucking sucks.

:)

It's not my code... it's our code. I don't take this discussion personal at all and I have no problem with anyone saying our code sucks (you know it oh too well). I believe when one points to code that sucks it's awesome and it's even more awesome to also provide/suggest a better alternative. And yes.. I do find the discussion around this issue awesome by itself, regardless of what ends up being pushed... simply because I know it eventually leads to a better code base.

This PR is not a defining moment in anyone's life. Like you, I have my opinions when it comes to code. Sometimes it's aligned with others, sometimes it's not, whatever it is, we always find a way to agree and move forward, simply because we all come with good intentions to improve things. You're the same... I don't expect you (or anyone working on elasticsearch) to be different - and that, my dear friend, is what makes you awesome. We discuss and we resolve thing by agreeing to agree, not by agreeing to disagree - that's a monumental pillar in our culture.

Bottom line, this PR introduces changes I'm -1 on. These changes I'm "-1'ing" are not blocking the actual bug fix. I suggest to remove the controversial part of the PR (which... again... does not block the bug fix) and open a separate issue to discuss the suggested "cleanup".

Lets see what others have to say about it... and act on additional feedback.

File: core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: yeah - no need to rename this

File: core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: can you elaborate on this I don't understand that problem necessarily?

File: core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java (hunk 4 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: I think we should reduce the scope here, remove the deprecation and open another issue to remove `SpawnModules` altogether. Elasticsearch core has gone through major refactorings to reduce the massive over-abstraction and over-engineering that we know as legacy code. Every project will get there and following the path we went down the last couple of weeks we should really make thinks clear and simple. Simple here means we need to add clean extension points without any modules depending on modules etc. you can extend a module and that's it. We should flatten out these extension points and don't allow for this recursive abstractions. It's hard enough to support one layer, multiple layers just make things harder and harder.

Comment (uboness) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: +1 on moving this to another issue and having the discussion there

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardsAllocatorModule.java [hunk 5]
```diff
@@ -64,8 +65,7 @@ protected void configure() {
             logger.warn("{} allocator has been removed in 2.0 using {} instead", EVEN_SHARD_COUNT_ALLOCATOR_KEY, BALANCED_ALLOCATOR_KEY);
             shardsAllocator = BalancedShardsAllocator.class;
         } else {
-            shardsAllocator = settings.getAsClass(TYPE_KEY, BalancedShardsAllocator.class,
-                    "org.elasticsearch.cluster.routing.allocation.allocator.", "Allocator");
+            throw new IllegalArgumentException("Unknown ShardsAllocator type [" + type + "]");
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardsAllocatorModule.java]: we can leave is a TODO  but I think we need a way to plug this in since there are quite some of them out there. I wonder if we can just have a `Map<String, Class<? extends ShardsAllocator>` here and a register method?

File: core/src/main/java/org/elasticsearch/repositories/RepositoryModule.java [hunk 6]
```diff
@@ -67,7 +66,11 @@ public RepositoryModule(RepositoryName repositoryName, Settings settings, Settin
      */
     @Override
     public Iterable<? extends Module> spawnModules() {
-        return ImmutableList.of(Modules.createModule(loadTypeModule(repositoryName.type(), "org.elasticsearch.repositories.", "RepositoryModule"), globalSettings));
+        Class<? extends Module> repoModuleClass = typesRegistry.type(repositoryName.type());
+        if (repoModuleClass == null) {
+            throw new IllegalArgumentException("Could not find repository type [" + repositoryName.getType() + "] for repository [" + repositoryName.getName() + "]");
+        }
+        return ImmutableList.of(Modules.createModule(repoModuleClass, globalSettings));
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/repositories/RepositoryModule.java]: just use `Collections.unmodifiableList()`

File: core/src/main/java/org/elasticsearch/transport/TransportModule.java [hunk 7]
```diff
@@ -48,6 +53,22 @@
     public TransportModule(Settings settings) {
         this.settings = settings;
         this.logger = Loggers.getLogger(getClass(), settings);
+        addTransport("local", LocalTransport.class);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TransportModule.java]: can we have constants here?

File: plugins/jvm-example/README.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+Example JVM Plugin for Elasticsearch
+==================================
+blah blah blah
```
Comment (s1monw) [inline plugins/jvm-example/README.md]: s/blah blah blah/Leniency is the root of all evil/

**Comment by s1monw** [comment]
I left some comments other than that LGTM. I'd really like to work on making all our internally plugged in mock things a `MockPlugin` too in a followup to make sure we can plug them in and at the same time we exercising our plugin system from core

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardsAllocatorModule.java (hunk 5 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardsAllocatorModule.java]: I created #12781 to do this as a follow up.

File: plugins/jvm-example/README.md (hunk 8 above)
Comment (rjernst) [inline plugins/jvm-example/README.md]: Nice.

**Comment by rjernst** [comment]
@uboness @s1monw I pushed new commits, removing the deprecation/rename and addressing other suggestions.

**Comment by s1monw** [comment]
LGTM

**Comment by s1monw** [comment]
this is a great step forward. Let's get rid of spawn modules altogether in a followup we need to flatten this hierarchy...

**Comment by s1monw** [comment]
oh and please squash these - it's pretty unreadable list of commits :)

**Comment by rjernst** [comment]
I created #12783 as a follow up.

File: core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java (hunk 4 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java]: +1 to break out deprecation to a separate issue

I see plenty of good fixes here, which everyone seems to agree on, and then also plenty of controversy around the deprecation, so I think it makes sense to split it off.

**Comment by rjernst** [comment]
I also created #12784 to finish the test refactoring work here.
